### DISCUSSION TITLE
GUIX/GUIX System on Lima

### DISCUSSION THREAD

**Discussion by msgilligan** [lima-vm/lima#1264] [General]
I am successfully running the [GUIX](https://guix.gnu.org) package manger in a Debian 11 VM installed with Lima.  The GUIX project produces both a package manger that works on existing Linux systems and also the [GUIX System](https://guix.gnu.org/en/manual/en/html_node/GNU-Distribution.html) which is a complete Linux distribution.

I would like to install the complete GUIX System in a macOS VM using Lima. I've done some preliminary research and found the well-written article [Guix System (GuixSD) VM on an M1 Mac](https://n8henrie.com/2022/10/guix-system-guixsd-vm-on-an-m1-mac/) which is really close to what I think I need. Along with its [preceding article](https://n8henrie.com/2022/09/linux-vms-on-my-m1-mac/) describes a process that can be summarized as:

1. Install an aarch64 Ubuntu guest using `libvirt`.
2. Install GUIX (package manager) on the Ubuntu guest
3. Build a GUIX System image using `guix system image`
4. `scp` the image back to macOS and further process with `qemu-img`
5. Install the aarch64 GUIX System guest using `libvirt`.

I'd like to replicate that procedure but with the following changes:

- Use Debian as the guess Linux for running the GUIX package manager
- Use Lima rather than `libvirt` for managing the macOS vms

Does anyone have any suggestions for me? Does the linked article and the 5 steps above look like the best approach?

**Comment by afbjorklund** [comment]
```
limactl start --name=guix template://debian
limactl shell guix sudo apt install -y guix
```

Then it can be used with a `lima` shell as usual:

`LIMA_INSTANCE=guix lima`

For instance, download the internet and say hello:

`guix install hello`


```console
anders@lima-guix:~$ guix install hello
guix install: warning: Consider running 'guix pull' followed by
'guix package -u' to get up-to-date packages and security updates.

The following package will be installed:
   hello 2.10

...
anders@lima-guix:~$ hello
Hello, world!
anders@lima-guix:~$ guix size hello
store item                                                       total    self
/gnu/store/fa6wj5bxkj5ll1d7292a70knmyl7a0cr-glibc-2.31              38.4    36.7  51.6%
/gnu/store/01b4w3m6mp55y531kyi1g8shh722kwqm-gcc-7.5.0-lib           71.0    32.6  45.8%
/gnu/store/mmhimfwmmidf09jw1plw3aw1g1zn2nkh-bash-static-5.0.16       1.6     1.6   2.3%
/gnu/store/a462kby1q51ndvxdv3b6p0rsixxrgx1h-hello-2.10              71.2     0.2   0.3%
total: 71.2 MiB
anders@lima-guix:~$ guix --version
guix (GNU Guix) 1.2.0
Copyright (C) 2020 the Guix authors
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

---

In order for "GUIX System" to work with Lima, it needs `cloud-init`.

* https://cloudinit.readthedocs.io/en/latest/

Lima uses this, along with some custom scripts for Ubuntu / Alpine,
to configure the VM. You can see the generated config in "cidata.iso"

* [docs/internal.md#cidataiso](https://github.com/lima-vm/lima/blob/master/docs/internal.md#cidataiso)

See this related discussion, about starting up NixOS with cloud-init:

* https://github.com/lima-vm/lima/discussions/430

Creating a GUIX "docker" <sic> container image would be another approach.

Then you could run that with `nerdctl`, instead of switching the VM OS ?

**Reply by msgilligan** [reply]
A quick Google found this:

https://lists.gnu.org/archive/html/guix-devel/2022-08/msg00187.html

It sounds like getting cloud-init integrated is no small task.

Maybe (at least as far as I use Lima) I should just stick with Guix-on-Debian.

(Related idea -- if there was a way to import a VM into Lima so that I could list/start/stop it with the same command that would be useful)

**Reply by afbjorklund** [reply]
In order for a VM to be "imported" to Lima, it would still have to create the user and install the guestagent and so on.
So that means that most of the cloud-init (or lima-init) tasks would need to have been performed on it, anyway... ?

You could probably write some simple shell wrapper to run the qemu commands, though. If not needing all of lima.

**Reply by msgilligan** [reply]
Maybe I'm trying to use Lima for something it wasn't intended for, but I really like how Lima sets up the VM with mounts, guest-agents, etc. I was also hoping I could use `limactl list` and the `~/.lima` directory to list/contain all my (regular, at least) VMs.

I'm new to Lima and to Guix, so I'm flying blind, here.  As it looks like integrating `cloud-init` into Guix System is unlikely to happen and `guix system image` is the way to create system images, I wonder how hard it would be to be able to use `guix system` to generate an image that would work with Lima?

**Reply by afbjorklund** [reply]
It should be possible to either run cloud-init (python) as-is, after patching in support for the "GUIX System" distribution.
Or perhaps port lima-init (shell, or possibly go in the future) from [alpine-lima](https://github.com/lima-vm/alpine-lima) to work with this special operating system...

Either way, you would have to supply the packaging (Guile?) and build the images - either "hard drive" or "live cd" as preferred.
It would also be theoretically possible to have lima work with "external" virtual machines, but that would miss out on features

Still think that running GUIX in a container should work ? It was one of the image types, and more in line with Lima ?

Unless you want to go all in on "Linux-libre" and "GNU Distribution", but that seems unlikely if already running a Mac

**Comment by msgilligan** [comment]
BTW, I am currently running a Lima NixOS VM (via [nixos-lima](https://github.com/nixos-lima/nixos-lima), which is implemented as @afbjorklund suggested above) with GUIX installed and it is working well.
